openshift_resources/resource.py

```diff
@@ -0,0 +1,112 @@
+from abc import ABCMeta
+from typing import Optional, Any
+
+from ocp_resources.resource import _get_client, _get_api_version
+from pydantic import BaseModel, Field
+from resources.io.k8s.apimachinery.pkg.apis.meta.v1 import ObjectMeta
+
+
+class BaseResource(BaseModel, metaclass=ABCMeta):
+    # fields included in the resource manifests; those fields could be "serialized"/"deserialized"
+    apiVersion: Optional[str]
+    kind: Optional[str]
+    metadata: Optional[ObjectMeta]
+
+    # excluded from the resource manifests
+    dyn_client: Optional[Any] = Field(exclude=True, repr=False, description="Represents the client of the resource")
+    api_group: Optional[str] = Field(exclude=True, repr=False, description="Represents the API group of the resource")
+
+    @property
+    def resource_client(self):
+        if not self.dyn_client:
+            self.set_client()
+
+        if not self.apiVersion:
+            self.discover_api_version()
+
+        if not self.kind:
+            self.discover_kind()
+
+        return self.dyn_client.resources.get(api_version=self.apiVersion, kind=self.kind)
+
+    def discover_api_version(self):
+        """
+        discover API version dynamically
+        :return:
+        """
+        if not self.apiVersion:
+            self.apiVersion = _get_api_version(
+                dyn_client=self.dyn_client, api_group=self.api_group, kind=self.kind
+            )
+
+    def discover_kind(self):
+        if not self.kind:
+            self.kind = self.__class__.__name__
+
+    def set_client(self, client=None, config_file=None, context=None):
+        if not client:
+            self.dyn_client = _get_client(
+                config_file=config_file, context=context
+            )
+        else:
+            self.dyn_client = client
+
+    def create(self):
```

Comment:
maybe  'create' can return a deserialized self as get does?

Reply:
no point - the return value is the same dict as the user pass as arguments